ucp/_libs/src/ucp_py_ucp_fxns.c

```diff
@@ -608,11 +631,43 @@ void *ucp_py_get_ep(char *ip, int listener_port)
                               send_handle);
     if (UCS_PTR_IS_ERR(request)) {
         fprintf(stderr, "unable to send UCX data message\n");
+	if (EP_CONN_FAILED == connection_status) {
+            ucp_request_cancel(ucp_py_ctx_head->ucp_worker, &request);
+            request_init(request);
+            ucp_request_free(request);
+            printf("connection failed. Retrying in %d seconds\n",
+                        ucp_py_ctx_head->connection_retry_period);
+            connection_status = EP_CONN_INIT;
+            sleep(ucp_py_ctx_head->connection_retry_period);
```

Comment:
This sleep is problematic from the perspective of an asynchronous framework (like Dask).  It's disruptive to block the main thread for more than several milliseconds.  Ideally making a connection would be asynchronous so that we could run other things in the background.

Reply:
Fair point. Would it be better if I returned `unreachable endpoint` error and then the coroutine which uses `get_endpoint` function can call say `asyncio.sleep` and retry later? Essentially leaving retry duration and policy to dask or some equivalent.

Reply:
Typically you return an unfinished thing immediately that might not have full connection information and then you have some asynchronous function to wait until connection has been achieved (which usually happens in the background. 

So you might do something like this:

```python
class Endpoint:
    def __init__(self, ...):
        ... # fast setup code only
        self._connected = asyncio.ensure_future(self.connect())

    async def connect(self, timeout=None):
        if timeout is not None:
            deadline = time() + timeout
        else:
            deadline = None
        while True:
            if try_to_connect():
                return self
            else:
                await gen.sleep(0.100)
                if time() > deadline:
                    raise TimeoutError("Failed to connect to %s after %f seconds" % (self.address, timeout))

    def __await__(self):  # this is the protocol for `await my_endpoint`
        return self._connected.__await__()  # I think this would work, but please don't trust me too much.
```

I think that with something like this we'll be able to do something like the following:

```python
ep = Endpoint(address)  # this is fast

# do other things

await ep

# now we can trust that ep is connected
```

Or, more commonly

```python
ep = await Endpoint(address)
```

Reply:
Awesome. This helps!